lib/Doctrine/DBAL/Cache/ResultCacheStatement.php

```diff
@@ -52,24 +52,29 @@ class ResultCacheStatement implements IteratorAggregate, ResultStatement, Result
     /** @var ResultStatement */
     private $statement;
 
+    /** @var array<string,mixed>|null */
+    private $fetchedData;
+
     /** @var array<int,array<string,mixed>>|null */
     private $data;
 
     /** @var int */
     private $defaultFetchMode = FetchMode::MIXED;
 
     /**
-     * @param string $cacheKey
-     * @param string $realKey
-     * @param int    $lifetime
+     * @param string                   $cacheKey
+     * @param string                   $realKey
+     * @param int                      $lifetime
+     * @param array<string,mixed>|null $fetchedData
```

Comment:
Should this new argument become mandatory in 3.0.x?

Reply:
I think, yes.
However, the class is renamed to CachingResult in 3.0

Reply:
Why don't we implement a proper fix in `3.0.x` first and then backport in a BC-way? Otherwise, it's not clear what the new intended logic is.

Reply:
I don't know how you releasing process is working. I consider this bug as a minor so it doesn't make sense to break backward compatibility.

Reply:
Let's at least make it clear in the parameter description that this parameter will be mandatory in `3.0`. Otherwise, it's not clear why fetching the data has to be implemented twice.

Reply:
I feel confused about how to do it all in 2.11.

I have opened the new https://github.com/doctrine/dbal/pull/4189 for 3.0.x version with BC Break and mandatory $fetchedData field.

Reply:
I'm not that well familiar with the design of the caching layer and I still don't get the design: instead of fetching the data from the result, why does the result now use the fetched data _and_ still can fetch from the result via `fetch*()` methods? Is this still necessary?

Reply:
The result cache is fetched in `ResultCacheStatement` only if no `fetchedData` is provided for constructor. It is done for backward compatibility - someone could use `ResultCacheStatement` before and it should continue working without the last parameter.

DBAL uses this class only in one place and it always sends `fetchedData`. So, result cache won't be fetched 2 times if DBAL is used ordinarily.

Reply:
> The result cache is fetched in `ResultCacheStatement` only if no `fetchedData` is provided to the constructor.

Let's then continue the discussion in https://github.com/doctrine/dbal/pull/4189 where BC is of no concern. The new class still depends on the `Result`.